core/controllers/async_download.py

```diff
@@ -0,0 +1,112 @@
+    rp_start: str | None = None,
+    rp_end: str | None = None,
+    outcome_categories: list[str] | None = None,
+):
+    """Download data, store file in S3 and send an email to the user with the download link."""
+
+    response = download(
+        file_format=file_format,
+        funds=funds,
+        organisations=organisations,
+        regions=regions,
+        rp_start=rp_start,
+        rp_end=rp_end,
+        outcome_categories=outcome_categories,
+    )
+
+    file_format_str: str = ""
+    if file_format == "xlsx":
+        file_format_str = "Microsoft Excel spreadsheet"
+    elif file_format == "json":
+        file_format_str = "JSON file"
+
+    file_size_kb = round(len(response.data) / 1024, 1)
+    if file_size_kb < 1024:
+        file_size_str: str = f"{round(file_size_kb, 1)} KB"
+    else:
+        file_size_str = f"{round(file_size_kb / 1024, 1)} MB"
+
+    # Upload the file to S3 and get presigned URL
+    file_obj = io.BytesIO(response.data)
+    bucket = Config.AWS_S3_BUCKET_FIND_DATA_FILES
+    find_service_base_url = Config.FIND_SERVICE_BASE_URL
+
+    if not bucket:
+        current_app.logger.error("AWS_S3_BUCKET_FIND_DATA_FILES is not set.")
+    elif not find_service_base_url:
+        current_app.logger.error("FIND_SERVICE_BASE_URL is not set.")
+    else:
+        file_name = f"{str(uuid.uuid4())}.{file_format}"
+        url = upload_file(file=file_obj, bucket=bucket, object_name=file_name, expiry=86400)
+
+        # email the URL to the user
+        send_email(
+            email_address=email_address,
+            download_url=url,
+            find_service_url=find_service_base_url + "/download",
+            file_format=file_format_str,
+            file_size_str=file_size_str,
+        )
+
+    return None
+
+
+def upload_file(file: IO | FileStorage, bucket: str, object_name: str, expiry: int) -> str:
+    """Upload a file to an S3 bucket, generate presigned url and return it
+    :param file: File to upload,
+    :param bucket: Bucket to upload to,
+    :param object_name: S3 object name (filename),
+    :param expiry: URL expiry time in seconds,
+
+    :return: Presigned URL
+    """
+
+    _S3_CLIENT.put_object(Bucket=bucket, Key=object_name, Body=file)
+
+    url = _S3_CLIENT.generate_presigned_url(
+        "get_object",
+        Params={"Bucket": bucket, "Key": object_name},
+        ExpiresIn=expiry,
+    )
+
+    return url
+
+
+def send_email(email_address: str, download_url: str, find_service_url: str, file_format: str, file_size_str: str):
+    """Send an email to the user with the download link.
+    :param email_address: user's email address,
+    :param download_url: download URL,
+    :param find_service_url: URL of the FIND service,
+    """
+
+    notifications_client = NotificationsAPIClient(current_app.config["NOTIFY_FIND_API_KEY"])
+
+    notifications_client.send_email_notification(
+        email_address=email_address,
+        template_id="62124580-5d5e-4975-ab84-76d14be2a9ad",
```

Comment:
I think this template ID should probably be on our Config object with a sensible name, eg `NOTIFY_TEMPLATE_FIND_DOWNLOAD_REPORT` - what do you think?

Reply:
Template ID has been added to Config object.